tracer/src/Datadog.Trace/ContinuousProfiler/ContextTracker.cs

```diff
@@ -0,0 +1,120 @@
+// <copyright file="ContextTracker.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Threading;
+using Datadog.Trace.ExtensionMethods;
+using Datadog.Trace.Logging;
+using Datadog.Trace.Util;
+
+namespace Datadog.Trace.ContinuousProfiler
+{
+    internal class ContextTracker
+    {
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(ContextTracker));
+
+        private readonly ProfilerStatus _status;
+        private readonly bool _isFlagSet;
+        private readonly ThreadLocal<IntPtr> _traceContextPtr;
+
+        public ContextTracker(ProfilerStatus status)
+        {
+            _status = status;
+            _isFlagSet = EnvironmentHelpers.GetEnvironmentVariable(ConfigurationKeys.CodeHotspotsEnabled)?.ToBoolean() ?? false;
+            _traceContextPtr = new ThreadLocal<IntPtr>();
+            Log.Information("CodeHotspots feature is {IsEnabled}.", _isFlagSet ? "enabled" : "disabled");
+        }
+
+        public bool IsEnabled
+        {
+            get
+            {
+                return _status.IsProfilerReady && _isFlagSet;
+            }
+        }
+
+        public void Set(ulong localRootSpanId, ulong spanId)
+        {
+            WriteToNative(new SpanContext(localRootSpanId, spanId));
+        }
+
+        public void Reset()
+        {
+            WriteToNative(SpanContext.Zero);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static void WriteContext(IntPtr ptr, in SpanContext ctx)
+        {
+            Marshal.StructureToPtr(ctx, ptr, false);
```

Comment:
I wouldn't assume that to be an atomic operation, no? There would then be a risk of tearing if there is a sample right in between writing the first field `LocalRootSpanId` and the second field `SpanId`. How is that handled?

Reply:
It's not an atomic operation.
We could handle it by having writing fence:
- Setting a byte indicating the profiler that an operation is progress
- writing the actual data
- Resetting the byte

We could also have a filtering in the profiler: if we see a callstack with that specific frame, we can just skip the IDs.
To be honest, we wanted to measure before applying courtermeasure (have a metric on the appearance of the frame)

How other profilers are handling that ?

Reply:
done in https://github.com/DataDog/dd-trace-dotnet/pull/2587/commits/8a653a12520a3cb263cd8aa042eb601c95d31ccb

Reply:
improved in https://github.com/DataDog/dd-trace-dotnet/pull/2587/commits/00e7b3907b0881925ff2e427a136ece4a0e6de21